### PR TITLE
error messages: include module inclusion errors for first class modules

### PR DESCRIPTION
**Pull request by Octachron**
Currently, the error message for subtyping errors on first class modules:
```ocaml
module type XY = sig  val x: int val y: int end
module type XYZ = sig include XY val z:int end
let convert x = (x: < f : < g : int -> (module XY); h:int >; i:int > :> <f: <g: int -> (module XYZ) > >)
```
stops abruptly at the frontier between the core level and the module level:
>```
>Error: Type < f : < g : int -> (module XYZ); h : int >; i : int >
>       is not a subtype of < f : < g : int -> (module XY) > >
>       Type (module XY) is not a subtype of (module XYZ)
>```

This PR proposes to expand the subtyping error message with the inclusion trace for first class modules:

>```
>Error: Type < f : < g : int -> (module XY); h : int >; i : int >
>       is not a subtype of < f : < g : int -> (module XYZ) > >
>       Type (module XY) is not a subtype of (module XYZ)
>       Modules do not match: XY is not included in XYZ
>       The value z is required but not provided
>```

This changes also makes it possible to explain that if
```
let convert x = (x: (module XYZ):>(module XY))
```

fails with

> Error: Type (module XYZ) is not a subtype of (module XY)

it is because the first-class module subtyping relationship requires that the two first-class modules share the same
memory representation:

>```
> Error: Type (module XYZ) is not a subtype of (module XY)
>       The two first-class module types differ by their runtime size.
>```

### CODE REVIEW

**Comment by gasche** [comment]
(cc @lthls)

**Comment by gasche** [comment]
In principle the change looks fine, I think this only needs a review for reasonableness and non-regression inside the type checker.

File: typing/ctype.ml
Comment (lthls) [inline typing/ctype.ml]: This could use a few more newlines

File: typing/ctype.ml
Comment (lthls) [inline typing/ctype.ml]: The type for `univar_pairs` is hard to read. I would be in favour of defining it as an alias near the top of the file, then using it both here and to annotate the `univar_pairs` global reference.

File: typing/ctype.ml
Comment (lthls) [inline typing/ctype.ml]: This part of the patch is tricky.
My understanding is that we're currently checking subtyping by trying unification then undoing it. If subtyping works, we still return constraints that will need to be re-unified later. If it doesn't work, we also return constraints that will be re-unified later, expecting that to fail but in a way that we can turn into a proper error message.
This patch tries to mix into these constraints some failures that are not unification failures but module typing errors. As far as I can tell, it only replaces unification constraints that would have failed anyway, so it doesn't change which programs are accepted but gives better errors for the ones which are not.
But I would prefer if someone with better understanding of this part of the code could confirm my intuition (in particular, I haven't checked that if `package_subtype` fails then unification of the package types must fail).

File: typing/errortrace.ml
Comment (lthls) [inline typing/errortrace.ml]: An extra blank line between the types wouldn't hurt

File: typing/errortrace.mli
Comment (lthls) [inline typing/errortrace.mli]: An extra blank line between the types wouldn't hurt

File: typing/includemod_errorprinter.ml
Comment (lthls) [inline typing/includemod_errorprinter.ml]: This part of the patch looks unnecessary

File: typing/includemod_errorprinter.ml
Comment (lthls) [inline typing/includemod_errorprinter.ml]: I don't think this comment is very useful.
I've been hit in the past by these weird coercions that are only relevant as part of a bigger structure coercion, and I think the solution should be to have two coercion types (module coercions and structure field coercions). For now, I think the code is fine but I'd rather have no comment than this unhelpful comment.

File: typing/includemod_errorprinter.ml
Comment (lthls) [inline typing/includemod_errorprinter.ml:120]: Could we have a comment explaining the `n < 0` case ?
My understanding is that before this patch, we couldn't reach the `n < 0` case because it would have caused another error earlier.
With this patch we also use this code in contexts where the condition can occur, but `n < 0` implies that there is a `Tcoerce_alias` or `Tcoerce_primitive` coercion and we will get a better message error using the `first_non_id` function below.

File: typing/includemod_errorprinter.ml
Comment (lthls) [inline typing/includemod_errorprinter.ml]: Maybe `positions` (plural) ?

File: testsuite/tests/typing-modules/firstclass.ml
Comment (lthls) [inline testsuite/tests/typing-modules/firstclass.ml]: I'm not sure if this is worth fixing, but the message points that `y` in `S5` occurs at the position of `mid` in `S4`, but the references to the package types earlier are in the reverse order, which could be a bit confusing.

**Review by lthls — APPROVED** [review]
I don't know enough of the type checker to vouch for the safety of this patch, but it looks reasonable.
My comments are a mix of style nitpicks and brain dumps of why I believe the code is correct; I don't think any of them should be blocking so I'm approving right now. Feel free to address my comments as you see fit.

File: typing/includemod_errorprinter.ml
Comment (Octachron) [reply typing/includemod_errorprinter.ml:120]: Indeed, the previous state of the code was using the invariant that it was only called on equivalent module types. Thus the coercion was necessarily a permutation of components.

Maybe

> (* if n < 0, this is not a permutation but a kind coercion, which will be covered in the `first_non_id` case *)

?

Comment (lthls) [reply typing/includemod_errorprinter.ml:120]: I don't know how it reads for someone familiar with this part of the code, but I spent some time wondering why we did not report this as an error before eventually managing to find that `not_fixpoint` runs before `first_non_id` and so we would get worse messages if we did not skip this case.
So I would prefer if you could manage to fit somewhere in the comment that `n < 0` is not a fixpoint (or rename `not_fixpoint` to `find_permutation`, this way your proposed comment becomes very clear).

Comment (Octachron) [reply typing/includemod_errorprinter.ml:120]: Hm, there are too many negative here: `n<0` is not a transposition (aka not not-a-fixpoint in the permutation). But this is indeed a sign that the names should be improved: `find_transposition` should be clearer (probably even for people that don't have the definition of transposition in mind?).

Comment (lthls) [reply typing/includemod_errorprinter.ml:120]: That works for me, indeed. Although the functions `transposition` and `transposition_under` now return a `change` that could be a transposition or not, so maybe a few more names need to be adapted.
If you don't want to spend too much time on this, you can also just adopt the comment in your earlier suggestion and leave further cleanup to an hypothetical future PR.

File: typing/ctype.ml
Comment (Octachron) [reply typing/ctype.ml]: The unification backtracking is specific to the first-class module path, where we need to unify the `with` constraint before comparing the two first-class module types.

But yes, the `Package_error _` constructor replaces unification constraints that will have failed later because during unification the first-module case calls once again  `package_subtype`, and in both direction.

However, thinking about it, it seems that a better implementation would be move the support of the first class module error to `unify_package_type` and the unification trace. The resulting code would probably be simpler.
I will have a look.

File: typing/includemod_errorprinter.ml
Comment (Octachron) [reply typing/includemod_errorprinter.ml:120]: I have changed few more names to better reflect that we are now extracting a small atomic change from any coercion rather than just a transposition from a permutation.

File: typing/includemod_errorprinter.ml
Comment (Octachron) [reply typing/includemod_errorprinter.ml]: I agree that `positions` sounds slightly better(even if this path should not happen).

File: testsuite/tests/typing-modules/firstclass.ml
Comment (Octachron) [reply testsuite/tests/typing-modules/firstclass.ml]: Keeping the package type and the associated item in the same order is very important. Fixed, thanks !

File: typing/ctype.ml
Comment (Octachron) [reply typing/ctype.ml]: The alternative implementation can be found here: https://github.com/ocaml/ocaml/commit/ae8a32b10bcc9fbf512776f84907de04476f9c5d#diff-89663ec75513e9fa325a3ba34185b45f278d35de3491e223bf2c0c49206a2a72 . Normally, I would propose this in a subsequent PR, but this implementation undoes a significant part of the changes from this PR (and is both smaller in code size and more complete in term of coverage). So it might be better to go with this version.

**Comment by gasche** [comment]
If I understand correctly, @Octachron did a substantial change just now, and we need a second review round of @lthls (or someone else) before we can merge.

**Comment by Octachron** [comment]
Indeed, after discussion with @lthls , it sounded better to re-review the PR after this change since it makes the PR substantially simpler while covering more case of first-class module related error messages.

**Comment by lthls** [comment]
Yes, I had already looked at the new commit and I plan to redo a review (I've tried to find a way to mark my previous review as stale, the closest I could fine was re-requesting a review, which also works).

File: typing/includemod_errorprinter.ml
Comment (lthls) [inline typing/includemod_errorprinter.ml]: Same as below, `position` could be plural (or `runtime components` could be singular).

File: typing/includemod_errorprinter.ml
Comment (lthls) [inline typing/includemod_errorprinter.ml:183]: The behaviour with `None` is changing a bit. The previous code had a comment saying "this should not happen", and there is some logic behind that (if we get an illegal permutation error we should not have an identity coercion), so I consider it's fine.
It does rely on identity coercions being systematically collapsed into `Tcoerce_none`, but I believe this is the case.

**Review by lthls — APPROVED** [review]
This looks more obviously correct than the previous version.

**Comment by shindere** [comment]
It seems the dependencies need to be updated, `make alldepend`.

Would have done it but was not sure how many commits you wanted to keep.

If you want to keep several commits and several of them change the
dependencies, it's IMHO nice if each commit includes the dependency
updates it induces.

File: typing/includemod_errorprinter.ml
Comment (Octachron) [reply typing/includemod_errorprinter.ml:183]: Indeed, this my usual hesitation between writing default error message for impossible case or failing early.

**Comment by Octachron** [comment]
I squashed and merged all commit together, since in this case there was conceptually a single change with one back-and-forth iteration.

**Pull request merged**
